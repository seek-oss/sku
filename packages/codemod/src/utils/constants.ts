type Codemod = {
  description: string;
  value: string;
};

export const CODEMODS = [
  {
    description:
      'Transform all webpack loadable imports to Vite loadable imports',
    value: 'transform-vite-loadable',
  },
  {
    description: 'Convert Jest tests to Vitest',
    value: 'jest-to-vitest',
  },
  /* [add-sku-codemod-generator: codemod] */
] as const satisfies Codemod[];

export type CodemodName = (typeof CODEMODS)[number]['value'];
