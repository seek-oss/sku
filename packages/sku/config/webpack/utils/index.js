const loaders = require('./loaders');
const { resolvePackage } = require('./resolvePackage');

module.exports = {
  ...loaders,
  resolvePackage,
  TYPESCRIPT: /\.(ts|tsx)$/,
  JAVASCRIPT: /\.js$/,
  LESS: /\.less$/,
  IMAGE: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
  SVG: /\.svg$/,
};
