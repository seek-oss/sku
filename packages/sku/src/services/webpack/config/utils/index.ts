import { extensions } from '../../../../config/fileResolutionExtensions.js';

export * from './loaders.js';
export { resolvePackage } from './resolvePackage.js';

export const TYPESCRIPT = new RegExp(`\.(${extensions.ts.join('|')})$`);
export const JAVASCRIPT = new RegExp(`\.(${extensions.js.join('|')})$`);
export const IMAGE = [
  /\.bmp$/,
  /\.gif$/,
  /\.jpe?g$/,
  /\.png$/,
  /\.webp$/,
  /\.avif$/,
];
export const SVG = /\.svg$/;
