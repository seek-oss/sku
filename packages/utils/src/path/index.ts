export { toPosixPath } from './toPosixPath.js';
