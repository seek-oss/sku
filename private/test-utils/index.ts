export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.ts';
export { dirContentsToObject } from './dirContentsToObject.ts';
export { makeStableHashes, makeStableViteHashes } from './skuConfig.ts';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.ts';
export { getPort } from './getPort.ts';
