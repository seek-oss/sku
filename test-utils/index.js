// @ts-check
export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.js';
export { getAppSnapshot } from './appSnapshot.js';
export { startAssetServer } from './assetServer.js';
export { dirContentsToObject } from './dirContentsToObject.js';
export { run, runSkuScriptInDir } from './process.ts';
export { makeStableHashes } from './skuConfig.js';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.js';
export { waitForUrls } from './waitForUrls.js';
