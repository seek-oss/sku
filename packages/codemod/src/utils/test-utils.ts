import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';
import { isValidTransform, type Transform } from './types.js';

type FileInfo = {
  source: string;
  path: string;
};

type InlineTestProps = {
  transform: Transform;
  input: FileInfo;
  expectedOutput: string;
};

type TestProps = {
  dirName: string;
  transformName: string;
  fixtureName?: string;
  testOptions?: { extension?: string };
};

const runInlineTest = async ({
  transform,
  input,
  expectedOutput,
}: InlineTestProps) => {
  if (!isValidTransform(transform)) {
    throw new Error('Invalid transform function provided');
  }

  const output = transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }

  expect(output.trim()).toEqual(expectedOutput.trim());

  return output;
};

export const runTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const testFixtureName = fixtureName || transformName;

  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${testFixtureName}.input.${testOptions.extension}`,
  );

  // Assumes transform is one level up from __tests__ directory
  const transformImport = import(path.join(dirName, '..', transformName));
  const sourceRead = readFile(inputPath, 'utf8');
  const expectedOutputRead = readFile(
    path.join(fixtureDir, `${testFixtureName}.output.${testOptions.extension}`),
    'utf8',
  );

  const [{ transform }, source, expectedOutput] = await Promise.all([
    transformImport,
    sourceRead,
    expectedOutputRead,
  ]);

  return runInlineTest({
    transform,
    input: {
      path: inputPath,
      source,
    },
    expectedOutput,
  });
};

export const runNoChangeTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${fixtureName}.unchanged.${testOptions.extension}`,
  );

  const sourceRead = readFile(inputPath, 'utf8');
  const transformImport = import(path.join(dirName, '..', transformName));

  const [{ transform }, source] = await Promise.all([
    transformImport,
    sourceRead,
  ]);
  const output = await transform(source);

  expect(output).toEqual(null);
};
