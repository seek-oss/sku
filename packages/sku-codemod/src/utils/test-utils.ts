import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';

type FileInfo = {
  source: string;
  path: string;
};

const runInlineTest = async (
  transform: (source: string) => string | false,
  input: FileInfo,
  expectedOutput: string,
) => {
  const output = await transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }
  const expectation = (o: string) =>
    expect(o.trim()).toEqual(expectedOutput.trim());
  expectation(output);
  return output;
};

export const runTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix?: string,
  testOptions: Record<string, unknown> = {},
) => {
  let localTestFilePrefix = testFilePrefix;
  if (!testFilePrefix) {
    localTestFilePrefix = transformName;
  }

  // Assumes transform is one level up from __tests__ directory
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${localTestFilePrefix}.input.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const expectedOutput = await readFile(
    path.join(
      fixtureDir,
      `${localTestFilePrefix}.output.${testOptions.extension}`,
    ),
    'utf8',
  );
  return runInlineTest(
    transform,
    {
      path: inputPath,
      source,
    },
    expectedOutput,
  );
};

export const runNoChangeTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix: string,
  testOptions: Record<string, unknown>,
) => {
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${testFilePrefix}.unchanged.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const output = await transform(source);
  expect(output).toEqual(false);
};
