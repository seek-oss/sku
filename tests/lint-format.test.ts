import {
  describe,
  test,
  beforeAll,
  afterAll,
  expect as globalExpect,
} from 'vitest';
import path from 'node:path';
import fs from 'node:fs/promises';
import dedent from 'dedent';
import { runSkuScriptInDir } from '@sku-private/test-utils';
import { createRequire } from 'node:module';
import { stripVTControlCharacters as stripAnsi } from 'node:util';

const require = createRequire(import.meta.url);

const appDirectory = path.dirname(
  require.resolve('@sku-fixtures/lint-format/sku.config.ts'),
);
const srcDirectory = path.join(appDirectory, 'src');
const testFile = (fileName: string) => path.join(srcDirectory, fileName);

const filesToLint: Record<string, string> = {
  'utils.test.ts': dedent/* ts */ `
    console.log('foo');

    it.only('should test something', () => {
      let foo = true;

      expect(foo).toBe(true);
    });\n`,
  'typescriptFile.ts': dedent/* ts */ `
    const foo: number = 'a string';
  `,
  'unformattedFile1.js': dedent/* js */ `
    import { something } from "with-double-quotes";

      const indented = 'indented';

    const foo = [
      "something \"really\" long",
      "that has to be moved",
      'to multiple lines',
      "so we can 'test' trailing commas"
    ];
  `,
};

const filesToFormat: Record<string, string> = {
  'importOrder1.ts': dedent/* ts */ `
    import './reset'; // side-effect imports should stay put

    import LocalComponent from './LocalComponent'; // sibling
    import Parent from '../parent'; // parents
    import someModule from 'some-module'; // external
    import vanillaStyles from './vanillaStyles.css'; // styles
    import distantParent from '../../../parent'; // parents
    import myself from '.'; // index
    import path from 'path'; //  built-in
    import utils from 'src/utils'; // internal
  `,
  'importOrder2.ts': dedent/* ts */ `
    import aStyle from './a.css';
    import bStyle from './b.css';
    import cStyle from '../c.css';

    import b from './b';
    import a from './a';
    import c from '../c';
  `,
  'fixableLintError.ts': dedent/* ts */ `
    const foo = () => {
      return 'foo';
    };
  `,
  'unformattedFile2.ts': dedent/* ts */ `
    import { something } from "with-double-quotes";

      const indented = 'indented';

    const foo = [
      "something \"really\" long",
      "that has to be moved",
      'to multiple lines',
      "so we can 'test' trailing commas"
    ];
  `,
};

globalExpect.addSnapshotSerializer({
  serialize: (val) => {
    const { stdout } = val;
    // Remove some logs that contain file paths that are unique to the machine
    const sanitizedStdout = stdout
      .split('\n')
      .filter((line: string) => !line.includes('sku/fixtures/lint-format'))
      .join('\n');

    return dedent`
      stdout: ${stripAnsi(sanitizedStdout)}
    `;
  },
  test: (val) => typeof val === 'object' && val.hasOwnProperty('stdout'),
});

describe('lint-format', () => {
  beforeAll(async () => {
    await fs.mkdir(srcDirectory, { recursive: true });
  });

  afterAll(async () => {
    await fs.rm(srcDirectory, { recursive: true, force: true });
  });

  describe('sku lint', () => {
    test.for(Object.keys(filesToLint))(
      'lint errors are reported: %s',
      async (fileName, { expect }) => {
        const filePath = testFile(fileName);
        await fs.writeFile(filePath, filesToLint[fileName]);

        await expect(
          runSkuScriptInDir('lint', appDirectory, { args: [filePath] }),
        ).rejects.toMatchSnapshot();
      },
    );
  });

  describe('sku format', () => {
    globalExpect.addSnapshotSerializer({
      serialize: (val) => val,
      test: (val) => typeof val === 'string',
    });

    test.for(Object.keys(filesToFormat))(
      'errors are fixed: %s',
      async (fileName, { expect }) => {
        const filePath = testFile(fileName);
        await fs.writeFile(filePath, filesToFormat[fileName]);

        await runSkuScriptInDir('format', appDirectory);

        const result = await fs.readFile(filePath, { encoding: 'utf-8' });

        expect(result).toMatchSnapshot();
      },
    );
  });
});
