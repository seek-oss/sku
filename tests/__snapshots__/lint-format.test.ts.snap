// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lint-format > sku format > errors are fixed: fixableLintError.ts 1`] = `
const foo = () => 'foo';

`;

exports[`lint-format > sku format > errors are fixed: importOrder1.ts 1`] = `
import './reset'; // side-effect imports should stay put

import path from 'path'; //  built-in

import someModule from 'some-module'; // external

import utils from 'src/utils'; // internal

import distantParent from '../../../parent'; // parents
import Parent from '../parent'; // parents

import LocalComponent from './LocalComponent'; // sibling

import myself from '.'; // index

import vanillaStyles from './vanillaStyles.css'; // styles

`;

exports[`lint-format > sku format > errors are fixed: importOrder2.ts 1`] = `
import c from '../c';

import a from './a';
import b from './b';

import aStyle from './a.css';
import bStyle from './b.css';
import cStyle from '../c.css';

`;

exports[`lint-format > sku format > errors are fixed: unformattedFile2.ts 1`] = `
import { something } from 'with-double-quotes';

const indented = 'indented';

const foo = [
  'something "really" long',
  'that has to be moved',
  'to multiple lines',
  "so we can 'test' trailing commas",
];

`;

exports[`lint-format > sku lint > lint errors are reported: typescriptFile.ts 1`] = `
stdout: Linting
Checking code with TypeScript compiler
src/typescriptFile.ts(1,7): error TS2322: Type 'string' is not assignable to type 'number'.
`;

exports[`lint-format > sku lint > lint errors are reported: unformattedFile1.js 1`] = `
stdout: Linting
Checking code with Prettier
src/unformattedFile1.js
To fix this issue, run 'pnpm run format'
`;

exports[`lint-format > sku lint > lint errors are reported: utils.test.ts 1`] = `
stdout: Linting
Checking code with TypeScript compiler
Checking code with Prettier
Checking code with ESLint

1:1  error  Unexpected console statement                    no-console
3:4  error  Unexpected focused test                         jest/no-focused-tests
4:7  error  'foo' is never reassigned. Use 'const' instead  prefer-const

âœ– 3 problems (3 errors, 0 warnings)
1 error and 0 warnings potentially fixable with the \`--fix\` option.
`;
